### ISSUE TITLE
[RFE] add support for Reader/Writer for chaining

### ISSUE THREAD

**Issue by ignatenkobrain** [alicemaz/rust-base64#20]
My use-case is pretty simple, I have file which contains base64-encoded, zlib-compressed JSON ;)

**Comment by ignatenkobrain** [comment]
I hope internally this could do something partially decoding base64 (since you can decode it by chunks), you don't need full string to be available.

**Comment by marshallpierce** [comment]
Are you picturing something like a `Base64BufReader` that wraps an underlying `BufRead`?

**Comment by ignatenkobrain** [comment]
yes, something like that. Unfortunately, I'm still in the middle of understanding how to chain FlateReader from flate2, so I can't come up with syntax how I would like to see it.

P.S. it's my first experience with chaining readers in Rust

**Comment by ignatenkobrain** [comment]
Also I'm not sure whether it should be BufRead or just Read

**Comment by ignatenkobrain** [comment]
```rust
impl EntitlementCertificate {
    pub fn from_file(file: &str) -> Self {
        let path = Path::new(file);

        let rdr = BufReader::new(File::open(path).unwrap());
        let b64iter = rdr.lines()
            .skip_while(|l| l.as_ref().unwrap() != ENTITLEMENT_DATA_HEADER)
            .skip(1)
            .take_while(|l| l.as_ref().unwrap() != ENTITLEMENT_DATA_FOOTER);
        let mut b64rdr = IterRead::new(b64iter);

        let mut data = String::new();
        b64rdr.read_to_string(&mut data).unwrap();
        let zlib_encoded = base64::decode(&data).unwrap();

        let decompressor = flate2::read::ZlibDecoder::new(&*zlib_encoded);

        serde_json::from_reader(decompressor).unwrap()
    }
}
```

that's what I've got so far in my project, as you can see, I have reading file, decompressing and deserializing working on top of Read.

**Comment by brianm** [comment]
I have 80% of an implementation of this internally at work. Basically is a Read and Write implementation loosely based on the golang implementation of the streaming encoder and decoder.

I'd be happy to extract the code and contribute it if there is interest.

**Comment by marshallpierce** [comment]
Sure, I'd be happy to take a look at that, if you feel like making a PR (even a rough one). I haven't forgotten about this feature; just got several things ahead of it on the to-do list. :)

BTW, thanks for JDBI!

**Comment by brianm** [comment]
Okay, got rid of the allocation at the expense of having to do extra reads sometimes (return less than full buffer), and I *think* need for access to config.pad. 

PR incoming.

**Comment by marshallpierce** [comment]
@brianm well, finally shipped encoding... decoding next? If you think your decoder is a good starting point I'm happy to run with it, or I can start from scratch.

**Comment by dignifiedquire** [comment]
@marshallpierce I have a working version of a streaming decoder here: https://github.com/dignifiedquire/pgp/blob/master/src/base64_decoder.rs if you want something as a starting point. It is modeled after the go implementation, but has some additional logic I needed, to be able to handle input ending for my use case.

It also is not very well tuned yet, in terms of how the buffers are used.

**Comment by marshallpierce** [comment]
Thanks; I'll look at it when I have some time!

**Comment by nwalfield** [comment]
I've implemented a streaming decoder.  Please see this MR: https://github.com/alicemaz/rust-base64/pull/106

**Comment by ggriffiniii** [comment]
In case anyone else stumbles here waiting for this feature. It's been implemented in the `radix64` crate. radix64::io has both an EncodeWriter and DecodeReader. https://docs.rs/radix64

**Comment by marshallpierce** [comment]
This has been implemented for some time now; see https://docs.rs/base64/latest/base64/write/index.html and https://docs.rs/base64/latest/base64/read/index.html.
